src/pkg/services/m365/api/contacts_pager.go

```diff
@@ -45,44 +49,57 @@ func (c Contacts) EnumerateContainers(
 		ByContactFolderIdString(baseContainerID).
 		ChildFolders()
 
-	for {
-		if el.Failure() != nil {
-			break
-		}
+	return &contactsFoldersPageCtrl{c.Stable, builder, options}
+}
 
-		resp, err := builder.Get(ctx, config)
-		if err != nil {
-			return graph.Stack(ctx, err)
-		}
+func (p *contactsFoldersPageCtrl) GetPage(
+	ctx context.Context,
+) (NextLinkValuer[models.ContactFolderable], error) {
+	resp, err := p.builder.Get(ctx, p.options)
+	return resp, graph.Stack(ctx, err).OrNil()
+}
 
-		for _, fold := range resp.GetValue() {
-			if el.Failure() != nil {
-				return el.Failure()
-			}
-
-			if err := graph.CheckIDNameAndParentFolderID(fold); err != nil {
-				errs.AddRecoverable(ctx, graph.Stack(ctx, err).Label(fault.LabelForceNoBackupCreation))
-				continue
-			}
-
-			fctx := clues.Add(
-				ctx,
-				"container_id", ptr.Val(fold.GetId()),
-				"container_display_name", ptr.Val(fold.GetDisplayName()))
-
-			temp := graph.NewCacheFolder(fold, nil, nil)
-			if err := fn(&temp); err != nil {
-				errs.AddRecoverable(ctx, graph.Stack(fctx, err).Label(fault.LabelForceNoBackupCreation))
-				continue
-			}
-		}
+func (p *contactsFoldersPageCtrl) SetNextLink(nextLink string) {
+	p.builder = users.NewItemContactFoldersItemChildFoldersRequestBuilder(nextLink, p.gs.Adapter())
+}
+
+func (p *contactsFoldersPageCtrl) ValidModTimes() bool {
+	return true
+}
+
+// EnumerateContainers iterates through all of the users current
+// contacts folders, transforming each to a graph.CacheFolder, and calling
+// fn(cf).
+// Contact folders are represented in their current state, and do
+// not contain historical data.
+func (c Contacts) EnumerateContainers(
```

Comment:
I feel like this is getting close to pushing too much functionality into what should ideally be a fairly thin wrapper around the Graph API. The original intent behind pulling out the Graph calls and placing them behind an interface was so we could mock them, allowing us to test a broader range of situations without having to manufacture state in the remote service

Granted, the original design is not particularly conducive to pulling the Graph calls out into a separate package, but we should be considerate of the original intent behind the refactoring effort

Reply:
Generally agreed.  Do I assume correctly that you've got your eye on the `fn` processor, in particular?  The rest of the func is more or less benign, imo.  If so, I'd be happy to follow up by splitting out the two parts (enumeration and transformation) to their respective layers.  But for this PR in particular, I'd prefer to keep the scope of changes where they are.

Reply:
yea, that's mostly the concern. The more complexity we push into functions like this, the more difficulty we'll have testing that these functions work as expected since they either require complex networking mock setups or special setups in the external service

As long as we prioritize splitting this to reduce the logic in functions like this is seems good

Reply:
PR up: https://github.com/alcionai/corso/pull/4346